### PR TITLE
fix: Handle types correctly in ResourceDictionary

### LINKED ISSUE #7044: `ResourceDictionary` shouldn't have a string key for `Style` without explicit `x:Name`
### Current behavior

```xml
    <Grid x:Name="grid">
        <Grid.Resources>
            <Button x:Name="MyButton" />
            <Style TargetType="StackPanel">
                <Setter Property="Margin" Value="4" />
            </Style>
            <Style TargetType="TextBlock" x:Name="MyStyle">
                <Setter Property="VerticalAlignment" Value="Center" />
                <Setter Property="Margin" Value="0,0,4,0"/>
            </Style>
        </Grid.Resources>
    </Grid>
```

```csharp
var resources = grid.Resources.ToList();
var a = resources[0];
var b = resources[1];
var c = resources[2];
```

All resources have a string Key.

### Expected behavior

Only `MyButton` and `MyStyle` have a string key.

UWP behavior:

image

### How to reproduce it (as minimally and precisely as possible)

↑

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

_No response_

### NuGet package version(s)

_No response_

### Affected platforms

_No response_

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #7044

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the current behavior?

Adding a resource with a System.Type key then retrieving it will result in a string being produced.

## What is the new behavior?

Correctly retrieve it as System.Type.

There was two issues here:

1. `ConvertKey(k.Key)`, this was resulting in an implicit conversion from string to `ResourceKey`, resulting in losing the information of the original `ResourceKey`.
2. Usage of `Type.GetType`, which only works in the current assembly or mscorlib (or System.Private.CoreLib).

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/SpecializedResourceDictionary.cs [hunk 1]
```diff
@@ -31,6 +31,7 @@ internal class SpecializedResourceDictionary
 		public readonly struct ResourceKey
 		{
 			public readonly string Key;
+			public readonly Type TypeKey;
 			public readonly bool IsType;
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/SpecializedResourceDictionary.cs:35]: We can certainly remove the `IsType` as a result now, as it is only used for type keys explicitly. That's remove a field to copy during by-value parameters.

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/SpecializedResourceDictionary.cs:35]: @jeromelaban Yeah sure. Will remove it. Do you want the HashCode to use `(TypeKey != null).GetHashCode()` (match current semantics), or `TypeKey.GetHashCode()` (probably better)?

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/SpecializedResourceDictionary.cs:35]: It's likely that `TypeKey?.GetHashCode() ?? 0` will be good enough.

**Review by jeromelaban — APPROVED** [review]
Thanks for your contribution!

**Pull request merged**